crates/ty_python_semantic/src/types/call/bind.rs

```diff
@@ -1085,23 +1085,152 @@ impl<'db> CallableBinding<'db> {
+
+        // Step 1: Check the result of the arity check which is done by `match_parameters`
+        match self.matching_overload_index() {
+            MatchingOverloadIndex::None => {
+                // If no candidate overloads remain from the arity check, we can stop here. We
+                // still perform type checking for non-overloaded function to provide better user
+                // experience.
+                if let [overload] = self.overloads.as_mut_slice() {
+                    overload.check_types(db, argument_types.as_ref(), argument_types.types());
+                }
+                return;
+            }
+            MatchingOverloadIndex::Single(index) => {
+                // If only one candidate overload remains, it is the winning match.
+                self.overloads.get_mut(index).unwrap().check_types(
+                    db,
+                    argument_types.as_ref(),
+                    argument_types.types(),
+                );
+                return;
+            }
+            MatchingOverloadIndex::Multiple(_) => {
+                // If two or more candidate overloads remain, proceed to step 2.
+            }
+        }
+
+        let pre_evaluation_snapshot = snapshot(self);
+
+        // Step 2: Evaluate each remaining overload as a regular (non-overloaded) call to determine
+        // whether it is compatible with the supplied argument list.
+        for (_, overload) in self.matching_overloads_mut() {
+            overload.check_types(db, argument_types.as_ref(), argument_types.types());
+        }
+
+        match self.matching_overload_index() {
+            MatchingOverloadIndex::None => {
+                // If all overloads result in errors, proceed to step 3.
+            }
+            MatchingOverloadIndex::Single(_) => {
+                // If only one overload evaluates without error, it is the winning match.
+                return;
+            }
+            MatchingOverloadIndex::Multiple(_) => {
+                // If two or more candidate overloads remain, proceed to step 4.
+                // TODO: Step 4 and Step 5 goes here...
+                // We're returning here because this shouldn't lead to argument type expansion.
+                return;
+            }
+        }
+
+        // Step 3: Perform "argument type expansion". Reference:
+        // https://typing.python.org/en/latest/spec/overload.html#argument-type-expansion
+
+        let mut expansions = argument_types.expand(db);
+
+        let Some(first_expansion) = expansions.next() else {
+            // Return early if there are no argument types to expand. This is especially useful to
+            // avoid restoring the bindings state.
+            return;
+        };
+
+        restore(self, pre_evaluation_snapshot);
+
+        for expanded_argument_lists in std::iter::once(first_expansion).chain(expansions) {
+            let mut return_types = vec![];
```

Comment:
Instead of building up a vec, and then only constructing and using the union if all expansions succeeded, you could build up in a `UnionBuilder` directly, and use a `bool` flag to tell you if one of the expansions fails and you should bail out.

Reply:
Yeah, I could do that. I did this to avoid the cost of [adding the type to the union](https://github.com/astral-sh/ruff/blob/a44db261387ffa44e7849b4d25e2ef3eed2524cf/crates/ty_python_semantic/src/types/builder.rs#L215-L410) in the cases where this expansion doesn't evaluate successfully and we need to move to the next argument type to expand or exit the expansion process because there are no argument type remaining.

Reply:
:+1: sounds good